### PR TITLE
httpc 301 redirect: Do not assert scheme ports are equal

### PR DESCRIPTION
**Pull request by rschlaikjer**
When handling 301 redirects from http -> https on Erlang 21.0.1, the
following error is encountered:

```
8> Options = [].
9> httpc:request(head, {"http://rhye.org", []}, Options, []).
{error,{shutdown,{{error,{badmatch,443}},
                  [{httpc_response,resolve_uri,7,
                                   [{file,"/usr/local/lib/erlang/lib/inets-7.0/src/http_client/httpc_response.erl"},
                                    {line,431}]},
                   {httpc_response,redirect,2,
                                   [{file,"/usr/local/lib/erlang/lib/inets-7.0/src/http_client/httpc_response.erl"},
                                    {line,396}]},
                   {httpc_handler,handle_response,1,
                                  [{file,"httpc_handler.erl"},{line,1052}]},
                   {httpc_handler,handle_info,2,
                                  [{file,"httpc_handler.erl"},{line,283}]},
                   {gen_server,try_dispatch,4,
                               [{file,"gen_server.erl"},{line,637}]},
                   {gen_server,handle_msg,6,
                               [{file,"gen_server.erl"},{line,711}]},
                   {proc_lib,init_p_do_apply,3,
                             [{file,"proc_lib.erl"},{line,249}]}]}}}
```

This seems to be caused by the following code in `resolve_uri`:
```erlang
resolve_uri(Scheme, Host, Port, Path, Query, URI, Map0) ->
    case maps:is_key(scheme, URI) of
        true ->
            Port = get_port(URI)
```

The value of `Port` passed in to `resolve_uri` here is 80, since the
original URL is http. However, since the redirected URL is https, the
`get_port` call returns 443, which is not equal to 80, and crashes.

Assigning to a new variable seems to fix redirects.

### CODE REVIEW

**Comment by rschlaikjer** [comment]
Trying to figure out a good way to add a test for this - it seems that the test suite in `httpc_SUITE.erl` is based around only running one server at a time for `http{,s}`, so a cross-scheme request like this doesn't fit in well. Would it be preferred to create a new test group for both http and https that spawns two servers, or to work the test into one of the existing groups somehow?

As far as I can tell, the following ought to test this behaviour once the dual-server setup is sorted:

```diff
diff --git a/lib/inets/test/httpc_SUITE.erl b/lib/inets/test/httpc_SUITE.erl
index d43e2cc179..094bcdd174 100644
--- a/lib/inets/test/httpc_SUITE.erl
+++ b/lib/inets/test/httpc_SUITE.erl
@@ -646,6 +647,26 @@ redirect_moved_permanently(Config) when is_list(Config) ->
        = httpc:request(post, {URL301, [],"text/plain", "foobar"},
                        [], []).
 %%-------------------------------------------------------------------------
+redirect_http_to_https() ->
+    [{doc, "The server SHOULD generate a Location header field in the response "
+      "containing a URI reference for the different URI.  The user agent MAY "
+      "use the Location field value for automatic redirection.  The server's "
+      "response payload usually contains a short hypertext note with a "
+      "hyperlink to the different URI(s)."}].
+redirect_http_to_https(Config) when is_list(Config) ->
+
+    URL302 = url(group_name(Config), "/302_with_ssl.html", Config),
+
+    {ok, {{_,200,_}, [_ | _], [_|_]}}
+       = httpc:request(get, {URL302, []}, [], []),
+
+    {ok, {{_,200,_}, [_ | _], []}}
+       = httpc:request(head, {URL302, []}, [], []),
+
+    {ok, {{_,200,_}, [_ | _], [_|_]}}
+       = httpc:request(post, {URL302, [],"text/plain", "foobar"},
+                       [], []).
+%%-------------------------------------------------------------------------
 redirect_found() ->
     [{doc, "The server SHOULD generate a Location header field in the response "
       "containing a URI reference for the different URI.  The user agent MAY "
@@ -2107,6 +2128,18 @@ handle_uri(_,"/302.html",Port, _,Socket,_) ->
        "Content-Length:" ++ integer_to_list(length(Body))
        ++ "\r\n\r\n" ++ Body;
 
+handle_uri(_,"/302_with_ssl.html",Port, _,Socket,_) ->
+    % Create a 302 redirect to a page with SSL enabled,
+    % regardless of the scheme of the previous request
+    NewUri = url_start(#sslsocket{}) ++
+       integer_to_list(Port) ++ "/dummy.html",
+    Body = "<HTML><BODY><a href=" ++ NewUri ++
+       ">New place</a></BODY></HTML>",
+    "HTTP/1.1 302 Found \r\n" ++
+       "Location:" ++ NewUri ++  "\r\n" ++
+       "Content-Length:" ++ integer_to_list(length(Body))
+       ++ "\r\n\r\n" ++ Body;
+
 handle_uri("HEAD","/303.html",Port,_,Socket,_) ->
     NewUri = url_start(Socket) ++
        integer_to_list(Port) ++ "/dummy.html",
```

**Comment by IngelaAndin** [comment]
I think it sounds better to make a new test group. Cramming something into a framework where it does not fit usually only makes the code harder to understand. Adding a new test SUITE can sometimes be an option too, but probably in this case a new group is enough.  You should be able to verify that your test case fails with the old code and passes with your addition.

File: lib/inets/src/http_client/httpc_response.erl [hunk 1]
```diff
@@ -425,11 +425,11 @@ resolve_uri(Scheme, Host, Port, Path, Query, URI) ->
 resolve_uri(Scheme, Host, Port, Path, Query, URI, Map0) ->
     case maps:is_key(scheme, URI) of
         true ->
-            Port = get_port(URI),
+            Port0 = get_port(URI),
```
[the remark this answers is not in the document]
Comment (rschlaikjer) [reply lib/inets/src/http_client/httpc_response.erl]: Good catch!

**Comment by rschlaikjer** [comment]
Verified that tests fail when patch is not applied, let me know if there's anything else I should change. Thank you for taking the time to review!

**Comment by peterdmv** [comment]
I think it would be cleaner if RelURI was not modified but the port is set together with the the scheme part.

```erlang
resolve_uri(Scheme, Host, Port, Path, Query, URI, Map0) ->
    case maps:get(scheme, URI, undefined) of
        undefined ->
            Port0 = get_port(Scheme, URI),
            Map = Map0#{scheme => Scheme,
                        port => Port0},
            resolve_authority(Host, Port, Path, Query, URI, Map);
        Scheme0 ->
            Port0 = get_port(Scheme0, URI),
            maybe_add_query(
              Map0#{scheme => Scheme0,
                    host => maps:get(host, URI),
                    port => Port0,
                    path => maps:get(path, URI)},
              URI)
    end.
```

```erlang
get_port(Scheme, URI) ->
    case maps:get(port, URI, undefined) of
        undefined ->
            get_default_port(Scheme);
        Port ->
            Port
    end.
```

```erlang
resolve_authority(Host, Port, Path, Query, RelURI, Map) ->
    case maps:get(host, RelURI, undefined) of
        undefined ->
            Map1 = Map#{host => Host,
                        port => Port},
            resolve_path(Path, Query, RelURI, Map1);
        Host0 ->
            maybe_add_query(
              Map#{host => Host0,
                   path => maps:get(path, RelURI)},
              RelURI)
    end.
```
Tests looks good!

**Comment by peterdmv** [comment]
All tests passed! Is it possible for you to rebase the PR to maint? We would like to have this fix in 21.1. If it's on master, it will be released in 22.

**Comment by rschlaikjer** [comment]
Excellent! I've rebased onto maint; guess I should have read the contributing guidelines more closely.

**Comment by peterdmv** [comment]
Thank you for your contribution!

**Comment by rschlaikjer** [comment]
Happy I could help, thank you for your time!

**Pull request merged**